### PR TITLE
update classes and tests to be compatible with new omnipay packages

### PR DESCRIPTION
**Pull request by digibeuk**
in line with the version numbers used bij omnipay I would suggest we make this available as the 3.0 branch.

~The minimum requirement for the package is also from now on php 7.1.~

See also : https://github.com/thephpleague/omnipay-common/releases/tag/v3.0.0

This should be throroughly tested.

### CODE REVIEW

**Comment by Feijs** [comment]
> in line with the version numbers used bij omnipay I would suggest we make this available as the 3.0 branch

Should it be merged to master then? If that is that case, a 1.0 branch should be made from the current master, so we can create updates and fixes untill the 3.0 branch is releasable and the transition to php ^7.1 has been made. 

Alternatively, I would suggest creating a 3.0 branch from master and directing this PR there

**Comment by digibeuk** [comment]
>Should it be merged to master then? If that is that case, a 1.0 branch should be made from the current master, so we can create updates and fixes untill the 3.0 branch is releasable and the transition to php ^7.1 has been made.

>Alternatively, I would suggest creating a 3.0 branch from master and directing this PR there
:+1:

File: src/Message/AbstractRequest.php [hunk 1]
```diff
@@ -2,43 +2,19 @@
...
 
-    /**
-     * @inheritdoc
-     */
-    public function __construct(ClientInterface $httpClient, HttpRequest $httpRequest)
-    {
-        parent::__construct($httpClient, $httpRequest);
-
-        // don't throw exceptions for 4xx errors
-        $this->httpClient->getEventDispatcher()->addListener(
```
Comment (Feijs) [inline src/Message/AbstractRequest.php]: How has this functionality been replaced? Is this a solved by the guzzle client update?

File: src/CurrencyAwareTrait.php [hunk 2]
```diff
@@ -2,30 +2,21 @@
...
+     * @param string $amount
      *
      * @return int
      */
     protected function toCurrencyMinorUnits($amount)
     {
-        return (int) round($amount * pow(10, $this->getCurrencyMinorUnitCount()));
+        $moneyParser = new DecimalMoneyParser(new ISOCurrencies());
+
+        return (int) $moneyParser->parse($amount, $this->getCurrency())->getAmount();
```
Comment (Feijs) [inline src/CurrencyAwareTrait.php]: This may throw a `ParserException` which is not exposed in the phpdoc

File: src/CurrencyAwareTrait.php [hunk 3]
```diff
@@ -2,30 +2,21 @@
...
-        return 2;
-    }
-
-    /**
-     * @param float|int $amount
+     * @param string $amount
      *
      * @return int
      */
     protected function toCurrencyMinorUnits($amount)
```
Comment (Feijs) [inline src/CurrencyAwareTrait.php]: Since php 7 is now required, this can become strict

File: src/Message/AbstractOrderRequest.php [hunk 4]
```diff
@@ -172,16 +174,22 @@ public function setWidgetOptions($widgetOptions)
...
-        ];
+        try {
+            $data = [
+                'order_amount' => $this->getAmountInteger(),
+                'order_tax_amount' => $this->toCurrencyMinorUnits((string) $this->getTaxAmount()),
+                'order_lines' => $this->getItemData($this->getItems()),
+                'purchase_currency' => $this->getCurrency(),
+                'purchase_country' => $this->getPurchaseCountry(),
+            ];
+        } catch (UnknownCurrencyException $exception) {
```
Comment (Feijs) [inline src/Message/AbstractOrderRequest.php]: Where does this exception originate, somewhere in `toCurrencyMinorUnits`?

File: src/Message/AbstractRequest.php [hunk 5]
```diff
@@ -232,47 +208,42 @@ public function setUsername($username)
...
+     * @return ResponseInterface
      */
     protected function sendRequest($method, $url, $data)
     {
         $options = (new AuthenticationRequestOptionProvider())->getOptions($this);
 
-        if (RequestInterface::GET === $method) {
-            return $this->httpClient->createRequest(
+        if ('GET' === $method) {
+            return $this->httpClient->request(
```
Comment (Feijs) [inline src/Message/AbstractRequest.php:231]: May throw `\Omnipay\Common\Http\Exception\NetworkException` and `\Omnipay\Common\Http\Exception\RequestException`

File: src/Message/RefundRequest.php [hunk 6]
```diff
@@ -30,15 +27,12 @@ public function getData()
     }
 
     /**
-     * {@inheritdoc}
-     *
-     * @throws RequestException
-     * @throws InvalidArgumentException
```
Comment (Feijs) [inline src/Message/RefundRequest.php]: The interface still does not declare these exceptions, why have they been removed here?

File: tests/Message/AbstractResponseTest.php [hunk 7]
```diff
@@ -1,77 +0,0 @@
-<?php
-
-namespace MyOnlineStore\Tests\Omnipay\KlarnaCheckout\Message;
-
-use MyOnlineStore\Omnipay\KlarnaCheckout\Message\AbstractResponse;
-use Omnipay\Tests\TestCase;
-
-class AbstractResponseTest extends TestCase
```
Comment (Feijs) [inline tests/Message/AbstractResponseTest.php]: Why has this test been removed? Looks like it is now covered by tests for **some** of it's implementations, but not all, which (imo) is an unclear practice.

File: src/Message/AbstractRequest.php (hunk 1 above)
Comment (digibeuk) [reply src/Message/AbstractRequest.php]: >A GuzzleHttp\Exception\ClientException is thrown for 400 level errors if the http_errors request option is set to true. This exception extends from GuzzleHttp\Exception\BadResponseException and GuzzleHttp\Exception\BadResponseException extends from GuzzleHttp\Exception\RequestException.

http://docs.guzzlephp.org/en/stable/request-options.html#http-errors

I thought it was automatically false, but true is the default, will set it to false!

**Comment by mattias-sanfridsson** [comment]
Is this still considered a WIP or is it ready to be merged?

**Comment by Feijs** [comment]
> Is this still considered a WIP or is it ready to be merged?

Looks like there are still a few unaddressed/unanswered questions, nor do I think it has been sufficiently tested in practice (at least not by me). So I would consider this WIP still. 

The author is currently unavailable, but I'm sure he can provide an update upon his return. And I'll do some testing in the meantime.

File: src/Message/AbstractRequest.php [hunk 8]
```diff
@@ -1,54 +1,55 @@
...
+            return $amount;
+        }
+
+        try {
+            $currency = new Currency($this->getCurrency());
+        } catch (\InvalidArgumentException $exception) {
+            $currency = new Currency('USD');
+        }
+
+        return new Money($amount, $currency);
```
Comment (Feijs) [inline src/Message/AbstractRequest.php]: This fails if amount is passed in decimal string format (which is currently the case). You'll probably need to use the DecimalMoneyParser here as well.

Same for `getTaxAmount`

File: src/Item.php [hunk 9]
```diff
@@ -7,7 +8,7 @@
     /**
      * @inheritDoc
      */
-    public function getMerchantData()
+    public function getMerchantData(): string
```
Comment (Feijs) [inline src/Item.php]: This can be null (or the return value should be cast to string)

File: src/Message/ItemDataTrait.php [hunk 10]
```diff
@@ -23,10 +21,10 @@ public function getItemData(ItemBag $items)
                 'type' => $item->getType(),
                 'name' => $item->getName(),
                 'quantity' => $item->getQuantity(),
-                'tax_rate' => (int) ($item->getTaxRate() * 100),
-                'total_amount' => $this->toCurrencyMinorUnits($item->getQuantity() * $item->getPrice()),
-                'total_tax_amount' => $this->toCurrencyMinorUnits($item->getTotalTaxAmount()),
-                'unit_price' => $this->toCurrencyMinorUnits($item->getPrice()),
+                'tax_rate' => (int) $item->getTaxRate(),
+                'total_amount' => $item->getQuantity() * $item->getPrice(),
+                'total_tax_amount' => (int) $item->getTotalTaxAmount(),
```
Comment (Feijs) [inline src/Message/ItemDataTrait.php]: `getTaxRate` and `getTotalTaxAmount` return floats, so can not just be cast to int like this. I suggest restoring the `toCurrencyMinorUnits`, of perhaps some alternative trait which can convert a float/string to a `Money` object so it can be used both here and in `AbstractRequest::getAmount`

File: src/CurrencyAwareTrait.php [hunk 11]
```diff
@@ -1,35 +1,23 @@
...
     {
-        return (int) round($amount * pow(10, $this->getCurrencyMinorUnitCount()));
-    }
+        $moneyParser = new DecimalMoneyParser(new ISOCurrencies());
 
-    /**
-     * @return string
-     */
-    abstract public function getCurrency();
+        return (int) $moneyParser->parse($money->getAmount(), $money->getCurrency());
```
Comment (Feijs) [inline src/CurrencyAwareTrait.php]: Money objects cannot be cast to int, `getAmount` is needed first

**Review by Feijs — CHANGES_REQUESTED** [review]
I propose to add some test-cases for the (string) input values that are currently used for prices/amounts

**Comment by digibeuk** [comment]
> Is this still considered a WIP or is it ready to be merged?

@mattias-persson it is still a WIP at the moment I am trying to find some time to get back at the task.

File: src/Message/AbstractOrderRequest.php (hunk 4 above)
Comment (digibeuk) [reply src/Message/AbstractOrderRequest.php]: IDK will remove it!

File: src/CurrencyAwareTrait.php [hunk 12]
```diff
@@ -4,20 +4,46 @@
...
+
+        try {
+            $currency = new Currency($this->getCurrency());
+        } catch (\InvalidArgumentException $exception) {
+            $currency = new Currency('USD');
+        }
+
+        $moneyParser = new DecimalMoneyParser(new ISOCurrencies());
+
+        return $moneyParser->parse((string) $amount, $currency);
```
Comment (Feijs) [inline src/CurrencyAwareTrait.php:35]: Can throw `Money\Exception\ParserException`

File: src/Message/AbstractRequest.php [hunk 13]
```diff
@@ -155,6 +136,11 @@ public function setItems($items)
             $items = new ItemBag($items);
         }
 
+
+        if ($items && !$items instanceof ItemBag) {
```
Comment (Feijs) [inline src/Message/AbstractRequest.php]: This is a duplicate from the block above?

File: src/Message/CaptureRequest.php [hunk 14]
```diff
@@ -3,11 +3,15 @@
...
 
+use MyOnlineStore\Omnipay\KlarnaCheckout\CurrencyAwareTrait;
 use Omnipay\Common\Exception\InvalidRequestException;
+use Omnipay\Common\Http\Exception\NetworkException;
+use Omnipay\Common\Http\Exception\RequestException;
 
 final class CaptureRequest extends AbstractRequest
 {
     use ItemDataTrait;
+    use CurrencyAwareTrait;
```
Comment (Feijs) [inline src/Message/CaptureRequest.php]: This is already part of the abstract base class

File: src/Message/ItemDataTrait.php [hunk 15]
```diff
@@ -17,18 +18,25 @@ public function getItemData(ItemBag $items): array
...
+
             $orderLines[] = [
                 'type' => $item->getType(),
                 'name' => $item->getName(),
                 'quantity' => $item->getQuantity(),
-                'tax_rate' => (int) $item->getTaxRate(),
-                'total_amount' => $item->getQuantity() * $item->getPrice(),
-                'total_tax_amount' => (int) $item->getTotalTaxAmount(),
-                'unit_price' => (int) $item->getPrice(),
+                'tax_rate' => null === $taxRate ? 0 : (int) $this->convertToMoney($taxRate)->getAmount(),
```
Comment (Feijs) [inline src/Message/ItemDataTrait.php]: `tax_rate` represent a percentage, are you sure converting this with currency context is the right way to go?

File: src/Message/RefundRequest.php [hunk 16]
```diff
@@ -3,24 +3,31 @@
...
 
+use MyOnlineStore\Omnipay\KlarnaCheckout\CurrencyAwareTrait;
 use Omnipay\Common\Exception\InvalidRequestException;
+use Omnipay\Common\Http\Exception\NetworkException;
+use Omnipay\Common\Http\Exception\RequestException;
 
 final class RefundRequest extends AbstractRequest
 {
     use ItemDataTrait;
+    use CurrencyAwareTrait;
```
Comment (Feijs) [inline src/Message/RefundRequest.php]: This is already part of the abstract base class

File: tests/Message/AuthorizeRequestTest.php [hunk 17]
```diff
@@ -261,7 +261,7 @@ public function testGetDataWithOptionsWillReturnCorrectData()
                 [
                     'locale' => 'nl_NL',
                     'amount' => '100.00',
-                    'tax_amount' => 2100,
+                    'tax_amount' => 21,
```
Comment (Feijs) [inline tests/Message/AuthorizeRequestTest.php]: Has the format the API expects for `tax_amount` changed from minor to mayor units? Otherwise this seems an incorrect expectation change

File: tests/Message/ItemDataTestTrait.php [hunk 18]
```diff
@@ -13,9 +13,9 @@ protected function getExpectedOrderLine(): array
             'type' => 'shipping_fee',
             'name' => 'item-name',
             'quantity' => 1,
-            'tax_rate' => 2003,
+            'tax_rate' => 200300,
             'total_amount' => 1000,
-            'total_tax_amount' => 200,
+            'total_tax_amount' => 20000,
```
Comment (Feijs) [inline tests/Message/ItemDataTestTrait.php]: A total tax amount which is 20 times the the total order amount seems an odd test case

File: tests/Message/ItemDataTestTrait.php [hunk 19]
```diff
@@ -13,9 +13,9 @@ protected function getExpectedOrderLine(): array
             'type' => 'shipping_fee',
             'name' => 'item-name',
             'quantity' => 1,
-            'tax_rate' => 2003,
+            'tax_rate' => 200300,
```
Comment (Feijs) [inline tests/Message/ItemDataTestTrait.php]: Why this change? This implies the input value has changed to have 4 decimals?

**Review by Feijs — CHANGES_REQUESTED** [review]
I've got an `omnipay-3` branch for the application to test this functionally. You can use that to check amounts are converted correctly etc.

File: tests/Message/ItemDataTestTrait.php [hunk 20]
```diff
@@ -13,7 +13,7 @@ protected function getExpectedOrderLine(): array
             'type' => 'shipping_fee',
             'name' => 'item-name',
             'quantity' => 1,
-            'tax_rate' => 200300,
+            'tax_rate' => 21,
```
Comment (Feijs) [inline tests/Message/ItemDataTestTrait.php]: The tax rate was explicitly defined with 2 decimals, to test/validate the case fixed in this PR:
https://github.com/MyOnlineStore/omnipay-klarna-checkout/pull/27/files

Though a comment would have been helpful. Perhaps you could add one?

**Review by Feijs — APPROVED** [review]
